### PR TITLE
feat: introduce first-class docker integration

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch:
- adds a `PWTEST_MODE=docker` mode for our own tests that runs
  all our tests inside docker container

Docker integration architecture:
- test code keeps running inside host environment
- browsers are being run inside local docker container

References #7439

### CODE REVIEW

**Review by dgozman — COMMENTED** [review]
- Let's remove fsTranslation business, as discussed offline.
- Let's think more about debugging story!
- I'd like someone who knows Docker more than I do to take a look :)

File: tests/browsercontext-proxy.spec.ts [hunk 1]
```diff
@@ -24,6 +24,8 @@ it.beforeEach(({ server }) => {
   });
 });
 
+it.beforeEach(({isDocker}) => it.skip(isDocker));
```
Comment (mxschmitt) [inline tests/browsercontext-proxy.spec.ts:27]:
```suggestion
test.skip(({isDocker}) => isDocker);
```

File: tests/chromium/tracing.spec.ts [hunk 2]
```diff
@@ -18,6 +18,8 @@ import { browserTest as it, expect } from '../config/browserTest';
 import fs from 'fs';
 import path from 'path';
 
+it.skip(({isDocker}) => !!isDocker);
```
Comment (mxschmitt) [inline tests/chromium/tracing.spec.ts:21]: nit: in all these calls you can remove the !!

File: tests/proxy.spec.ts [hunk 3]
```diff
@@ -18,6 +18,8 @@ import { playwrightTest as it, expect } from './config/browserTest';
 import socks from 'socksv5';
 import net from 'net';
 
+it.beforeEach(({isDocker}) => it.skip(isDocker));
```
Comment (mxschmitt) [inline tests/proxy.spec.ts:21]:
```suggestion
test.skip(({isDocker}) => isDocker);
```

File: tests/trace-viewer/trace-viewer.spec.ts [hunk 4]
```diff
@@ -113,6 +113,8 @@ const test = playwrightTest.extend<{ showTraceViewer: (trace: string) => Promise
   }
 });
 
+test.beforeEach(({isDocker}) => test.skip(isDocker));
```
Comment (mxschmitt) [inline tests/trace-viewer/trace-viewer.spec.ts:116]:
```suggestion
test.skip(({isDocker}) => isDocker);
```

File: src/client/browser.ts [hunk 5]
```diff
@@ -48,7 +49,9 @@ export class Browser extends ChannelOwner<channels.BrowserChannel, channels.Brow
     this._closedPromise = new Promise(f => this.once(Events.Browser.Disconnected, f));
   }
 
-  _setBrowserType(browserType: BrowserType) {
+  _connectToParent({browserType, logger, remoteType}: {browserType: BrowserType, logger: Logger|undefined, remoteType: RemoteType}) {
```
Comment (dgozman) [inline src/client/browser.ts:52]: I'd make it 3 separate arguments.

File: src/client/browserType.ts [hunk 6]
```diff
@@ -190,9 +195,11 @@ export class BrowserType extends ChannelOwner<channels.BrowserTypeChannel, chann
           }
 
           const browser = Browser.from(playwright._initializer.preLaunchedBrowser!);
-          browser._logger = logger;
-          browser._remoteType = 'owns-connection';
-          browser._setBrowserType((playwright as any)[browser._name]);
+          browser._connectToParent({
+            logger,
+            remoteType: this._isRemote ? 'uses-connection' : 'owns-connection',
```
Comment (dgozman) [inline src/client/browserType.ts:200]: This should always be `'owns-connection'`.

File: src/remote/playwrightClient.ts [hunk 7]
```diff
@@ -58,6 +58,9 @@ export class PlaywrightClient {
 
   constructor(playwright: Playwright, ws: WebSocket) {
     this._playwright = playwright;
+    this._playwright.webkit._isRemote = true;
```
Comment (dgozman) [inline src/remote/playwrightClient.ts:61]: I now think that we can do `Connection.isRemote` instead, so we will specify whether the connection is remote or local. WDYT?

File: src/test/playwrightDocker.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+  }
+
+  toDockerPath(hostPath: string): string {
+    const mount = this._mounts.find(mount => hostPath.startsWith(mount.hostPath));
+    if (!mount)
+      return hostPath;
+    return mount.dockerPath + hostPath.substring(mount.hostPath.length);
+  }
+
+  fromDockerPath(dockerPath: string): string {
```
Comment (dgozman) [inline src/test/playwrightDocker.ts:53]: This one seems unused?

File: src/test/playwrightDocker.ts [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+import { setStackTranslator } from '../utils/stackTrace';
+import { PlaywrightClient } from '../remote/playwrightClient';
+
+type DockerMount = {
+  hostPath: string,
+  dockerPath: string,
+};
+
+export class PlaywrightDocker {
+  private _client: any;
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/test/playwrightDocker.ts:36]: That's a bad signal. Could you share the errors please? Why can't this just be `_client: PlaywrightClient`?

File: tests/trace-viewer/trace-viewer.spec.ts [hunk 10]
```diff
@@ -113,6 +113,8 @@ const test = playwrightTest.extend<{ showTraceViewer: (trace: string) => Promise
   }
 });
 
+test.skip(({isDocker}) => isDocker);
```
Comment (dgozman) [inline tests/trace-viewer/trace-viewer.spec.ts:116]: I am curious why doesn't this file work in docker?

File: types/test.d.ts [hunk 11]
```diff
@@ -2288,6 +2288,10 @@ export interface PlaywrightWorkerOptions {
    * `devtools` option is `true`.
    */
   headless: boolean | undefined;
+  /**
+   * Wheather to run browsers inside docker container.
```
Comment (dgozman) [inline types/test.d.ts:2292]: typo: wheather

File: .github/workflows/tests_primary.yml [hunk 12]
```diff
@@ -66,3 +66,34 @@ jobs:
     - run: npm run build
     - run: node lib/cli/cli install --with-deps
     - run: npm run ttest
+
+  test_docker_mode:
+    name: Docker Mode ${{ matrix.os }} (${{ matrix.browser }})
```
Comment (mxschmitt) [inline .github/workflows/tests_primary.yml:71]: Seems like you deleted the wrong GitHub Action workflow file.

Comment (aslushnikov) [reply .github/workflows/tests_primary.yml:71]: reverted. The original idea was to cut on our CI usage, but turns out we have to test both `root` and `pwuser` usecases.

File: src/test/index.ts [hunk 13]
```diff
@@ -37,6 +37,7 @@ export const test = _baseTest.extend<TestFixtures, WorkerAndFileFixtures>({
   defaultBrowserType: [ 'chromium', { scope: 'worker' } ],
   browserName: [ ({ defaultBrowserType }, use) => use(defaultBrowserType), { scope: 'worker' } ],
   playwright: [ require('../inprocess'), { scope: 'worker' } ],
+
```
Comment (mxschmitt) [inline src/test/index.ts:40]:
```suggestion
```

**Comment by aslushnikov** [comment]
Closing in favor of https://github.com/microsoft/playwright/pull/8885

**Pull request closed without merging**